### PR TITLE
Add transformers to change detection

### PR DESCRIPTION
**Pull request by vsavkin**
- Even though the PR does not change the binding syntax, the looking up of transformers happens through the transformer repo.

### CODE REVIEW

File: modules/angular2/src/change_detection/transformers/transformer_registry.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import {isBlank, isPresent, BaseException, CONST} from 'angular2/src/facade/lang';
+
+export class TransformerRegistry {
+  config;
+
+  constructor(config){
+    this.config = config;
+  }
+
+  get(type:string, obj) {
```
Comment (mhevery) [inline modules/angular2/src/change_detection/transformers/transformer_registry.js]: Add return type annotation

File: modules/angular2/change_detection.js [hunk 2]
```diff
@@ -23,15 +28,30 @@ export class ChangeDetection {
   }
 }
 
+export var defaultTransformers = {
+  "[]" : [
+    {
+      "supports" : ArrayChanges.supportsObj,
+      "transformer" : () => new ArrayChanges()
```
Comment (mhevery) [inline modules/angular2/change_detection.js]: use same static tear of as in supports

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 [hunk 3]
```diff
@@ -162,14 +164,18 @@ if (${CHANGES_LOCAL} && ${CHANGES_LOCAL}.length > 0) {
...
+function transformerCheckTemplate(context:string, transformer:string,
+                                  value:string, change:string, addRecord:string, notify:string):string{
   return `
-${CHANGE_LOCAL} = ${UTIL}.structuralCheck(${field}, ${context});
-if (${CHANGE_LOCAL}) {
-  ${CHANGES_LOCAL} = ${UTIL}.addRecord(${CHANGES_LOCAL},
-    ${UTIL}.changeRecord(${PROTOS_ACCESSOR}[${selfIndex}].bindingMemento, ${CHANGE_LOCAL}));
-  ${field} = ${CHANGE_LOCAL}.currentValue;
+if (${transformer} === ${UTIL}.unitialized() || !${transformer}.supports(${context})) {
+  ${transformer} = ${TRANSFORMER_REGISTRY_ACCESSOR}.get('[]', ${context});
```
Comment (mhevery) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: `onDestroy`

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 [hunk 4]
```diff
@@ -236,6 +242,10 @@ export class ChangeDetectorJITGenerator {
   changeNames:List<String>;
   fieldNames:List<String>;
 
+  transformerNames:List<String>;
+  transformerStateNames:List<String>;
```
Comment (mhevery) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: remove

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 [hunk 5]
```diff
@@ -236,6 +242,10 @@ export class ChangeDetectorJITGenerator {
...
+  transformerPrevContextNames:List<String>;
```
Comment (mhevery) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: remove

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 [hunk 6]
```diff
@@ -295,17 +318,24 @@ export class ChangeDetectorJITGenerator {
   }
 
   genRecord(r:ProtoRecord):string {
-    if (r.mode == RECORD_TYPE_STRUCTURAL_CHECK) {
-      return this.getStructuralCheck(r);
+    if (r.mode === RECORD_TYPE_STRUCTURAL_CHECK) {
+      return this.getTransformerCheck(r);
```
Comment (mhevery) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: `genTransformerCheck`

File: modules/angular2/src/directives/foreach.js [hunk 7]
```diff
@@ -7,21 +6,19 @@ import {ListWrapper} from 'angular2/src/facade/collection';
 
 @Viewport({
   selector: '[foreach][in]',
-  lifecycle: [onChange],
   bind: {
-    'in': 'iterable[]'
+    'in': 'iterableChanges[]'
```
Comment (mhevery) [inline modules/angular2/src/directives/foreach.js]: `iterableChanges | observable | iterableDiffer`

File: modules/angular2/change_detection.js [hunk 8]
```diff
@@ -23,15 +28,30 @@ export class ChangeDetection {
   }
 }
 
+export var defaultTransformers = {
```
Comment (mhevery) [inline modules/angular2/change_detection.js]: Rename `transformers` as `pipes`

**Comment by vicb** [comment]
I've only checked the code quickly but if I understand correctly I don't like the "pipe" terminology. When you put something in a "pipe" you should get the same thing out of the "pipe". Doesn't the code "pipe" to a "transformer" ? If this is the case "transformer" would be a better "term" than "pipe".

**Comment by mhevery** [comment]
@vicb I am on the fence between the Pipe vs Transformer. We already have pipes in the syntax `{{ items | filter: predicate | order}}` 

Transformer makes more sense for `foreach` when binding and pipes make more sense for `{{}}`.

Do you have an alternative proposal?

**Pull request merged**